src/analysis/special_functions/exp_log.lean

```diff
@@ -389,107 +421,149 @@ end
-  (hf : continuous f) : continuous (λa, log (f a)) :=
-show continuous ((log ∘ @subtype.val ℝ (λr, 0 < r)) ∘ λa, ⟨f a, h a⟩),
-  from continuous_log'.comp (continuous_subtype_mk _ hf)
+lemma continuous_log' : continuous (λ x : {x : ℝ // 0 < x}, log x) :=
+continuous_on_iff_continuous_restrict.1 $ continuous_on_log.mono $ λ x hx, ne_of_gt hx
 
-end prove_log_is_continuous
+lemma continuous_at_log (hx : x ≠ 0) : continuous_at log x :=
+(continuous_on_log x hx).continuous_at $ mem_nhds_sets is_open_compl_singleton hx
+
+@[simp] lemma continuous_at_log_iff : continuous_at log x ↔ x ≠ 0 :=
+begin
+  refine ⟨_, continuous_at_log⟩,
+  rintros h rfl,
+  exact not_tendsto_nhds_of_tendsto_at_bot tendsto_log_nhds_within_zero _
+    (h.tendsto.mono_left inf_le_left)
+end
 
 lemma has_deriv_at_log_of_pos (hx : 0 < x) : has_deriv_at log x⁻¹ x :=
 have has_deriv_at log (exp $ log x)⁻¹ x,
-from (has_deriv_at_exp $ log x).of_local_left_inverse (continuous_at_log hx)
-  (ne_of_gt $ exp_pos _) $ eventually.mono (mem_nhds_sets is_open_Ioi hx) @exp_log,
+from (has_deriv_at_exp $ log x).of_local_left_inverse (continuous_at_log hx.ne')
+  (ne_of_gt $ exp_pos _) $ eventually.mono (lt_mem_nhds hx) @exp_log,
 by rwa [exp_log hx] at this
 
 lemma has_deriv_at_log (hx : x ≠ 0) : has_deriv_at log x⁻¹ x :=
 begin
-  by_cases h : 0 < x, { exact has_deriv_at_log_of_pos h },
-  push_neg at h,
-  convert ((has_deriv_at_log_of_pos (neg_pos.mpr (lt_of_le_of_ne h hx)))
-    .comp x (has_deriv_at_id x).neg),
-  { ext y, exact (log_neg_eq_log y).symm },
-  { field_simp [hx] }
+  cases hx.lt_or_lt with hx hx,
+  { convert (has_deriv_at_log_of_pos (neg_pos.mpr hx)).comp x (has_deriv_at_neg x),
+    { ext y, exact (log_neg_eq_log y).symm },
+    { field_simp [hx.ne] } },
+  { exact has_deriv_at_log_of_pos hx }
 end
 
+lemma differentiable_at_log (hx : x ≠ 0) : differentiable_at ℝ log x :=
+(has_deriv_at_log hx).differentiable_at
+
+lemma differentiable_on_log : differentiable_on ℝ log {0}ᶜ :=
+λ x hx, (differentiable_at_log hx).differentiable_within_at
+
+@[simp] lemma differentiable_at_log_iff : differentiable_at ℝ log x ↔ x ≠ 0 :=
+⟨λ h, continuous_at_log_iff.1 h.continuous_at, differentiable_at_log⟩
+
+lemma deriv_log (x : ℝ) : deriv log x = x⁻¹ :=
+if hx : x = 0 then
+  by rw [deriv_zero_of_not_differentiable_at (mt differentiable_at_log_iff.1 (not_not.2 hx)), hx,
+    inv_zero]
+else (has_deriv_at_log hx).deriv
+
+@[simp] lemma deriv_log' : deriv log = has_inv.inv := funext deriv_log
+
 lemma measurable_log : measurable log :=
 measurable_of_measurable_on_compl_singleton 0 $ continuous.measurable $
-  continuous_iff_continuous_at.2 $ λ x, (real.has_deriv_at_log x.2).continuous_at.comp
-    continuous_at_subtype_coe
+  continuous_on_iff_continuous_restrict.1 continuous_on_log
+
+lemma times_cont_diff_on_log {n : with_top ℕ} : times_cont_diff_on ℝ n log {0}ᶜ :=
+begin
+  suffices : times_cont_diff_on ℝ ⊤ log {0}ᶜ, from this.of_le le_top, clear n,
```

Comment:
```suggestion
  suffices : times_cont_diff_on ℝ ⊤ log {0}ᶜ, from this.of_le le_top,
```
(sorry if this has some purpose I failed to notice)

Reply:
This was a note to the reader (and to myself while writing the proof): we no longer use the original `n`.